src/core/unif.ml

```diff
@@ -68,12 +68,23 @@ let try_rules : ctxt -> term -> term -> constr list option = fun ctx s t ->
   let exception No_match in
   let open Unif_rule in
   try
+    let a_s,constrs_s = Infer.infer ctx s in
+    let a_t,constrs_t = Infer.infer ctx t in
```

Comment:
I don't understand why you need to do type inference to apply unif_rules.

Reply:
As already mentioned, in fact, I don't understand why you need to change tree_walk by adding an extra argument. If those two things are an attempt to fix the problem with `set unif_rule fst $p ≡ $f ↪ $p ≡ pair $f $s`, they should be removed as I think this is a bug in the handling of unif_rules.

Reply:
I will removed everything in the PR related to set unif_rule and we will think about it in another PR later

It was indeed an attempt (but not sufficient) to fill an array in tree_walk with the correct type instead of Kind, but this has to be investigated further

Reply:
Ok. And you should open another issue on this topic: the application of unification rules does not create well typed metavariables. @gabrielhdt

Reply:
I see that you have already done that. Thanks